integration/spark/spark3/src/main/java/io/openlineage/spark3/agent/lifecycle/plan/DataSourceV2ScanRelationInputDatasetBuilder.java

```diff
@@ -29,19 +32,37 @@ public DataSourceV2ScanRelationInputDatasetBuilder(
     this.factory = factory;
   }
 
+  @Override
+  public boolean isDefinedAt(SparkListenerEvent event) {
+    /* (2024-08-15) I implemented this method because it deferred to super.isDefinedAt(SparkListenerEvent),
+    which only looks at SparkListenerJobStart and SparkListenerSQLExecutionStart. However, due to
+    https://github.com/OpenLineage/OpenLineage/issues/2935, openlineage-spark was not adding the actual
+    input dataset, and debugging the code eventually led me to this point. For some reason, the
+    SparkListenerSQLExecutionStart event didn't have the necessary details to create an ExecutionContext.
+    Therefore, this code was never evaluated for the SparkListenerSQLExecutionStart. I'm keeping the
+    SparkListenerSQLExecutionStart defined here, just in case there is a code path that uses it.*/
+    return event instanceof SparkListenerSQLExecutionStart
+        || event instanceof SparkListenerSQLExecutionEnd;
+  }
+
   @Override
   public boolean isDefinedAtLogicalPlan(LogicalPlan logicalPlan) {
     return logicalPlan instanceof DataSourceV2ScanRelation;
   }
 
   @Override
   public List<OpenLineage.InputDataset> apply(DataSourceV2ScanRelation scanRelation) {
-    DataSourceV2Relation relation = (scanRelation).relation();
+    DataSourceV2Relation relation = scanRelation.relation();
     OpenLineage.DatasetFacetsBuilder datasetFacetsBuilder =
         context.getOpenLineage().newDatasetFacetsBuilder();
 
     DatasetVersionDatasetFacetUtils.includeDatasetVersion(context, datasetFacetsBuilder, relation);
```

Comment:
This builder is called on start only bcz of this facet. The corner case is a job reading and writing to the same dataset. In this scenario, including input at the job end can lead to collecting incorrect version facet. Can we include version on start only?

Reply:
Can you provide a test (in Java / Python / Scala) that replicates the issue you're describing?